template/interpolate/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,162 @@
+	keyVaultBaseUri, err := GetKeyVaultBaseUrlFromID(context.Background(), c.VaultsClient, keyVaultId)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := c.ManagementClient.GetSecret(context.Background(), keyVaultBaseUri, key, "")
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.Value, nil
+}
+
+func newVaultsClients() (*keyvault.VaultsClient, *keyvaultmgmt.BaseClient, error) {
+	builder := &authentication.Builder{
+		SupportsClientCertAuth:   true,
+		SupportsClientSecretAuth: true,
+		SupportsAzureCliToken:    true,
+	}
+
+	config, err := builder.Build()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	env, err := authentication.DetermineEnvironment(config.Environment)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	oauthConfig, err := config.BuildOAuthConfig(env.ActiveDirectoryEndpoint)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// OAuthConfigForTenant returns a pointer, which can be nil.
+	if oauthConfig == nil {
+		return nil, nil, fmt.Errorf("Unable to configure OAuthConfig for tenant %s", config.TenantID)
+	}
+
+	sender := sender.BuildSender("AzureRM")
+	keyVaultAuth, err := config.GetAuthorizationToken(sender, oauthConfig, env.TokenAudience)
+	if err != nil {
+		return nil, nil, err
+	}
+	endpoint := env.ResourceManagerEndpoint
+
+	VaultsClient := keyvault.NewVaultsClientWithBaseURI(endpoint, config.SubscriptionID)
+	VaultsClient.Authorizer = keyVaultAuth
+
+	ManagementClient := keyvaultmgmt.New()
+	MgmtAuthorizer := config.BearerAuthorizerCallback(sender, oauthConfig)
+	ManagementClient.Client.Authorizer = MgmtAuthorizer
+
+	return &VaultsClient, &ManagementClient, nil
+}
+
+func GetKeyVaultBaseUrlFromID(ctx context.Context, client *keyvault.VaultsClient, keyVaultId string) (string, error) {
+	resourceGroup, vaultName, err := getKeyVaultResourceGroupAndName(keyVaultId)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := client.Get(ctx, resourceGroup, vaultName)
+	if err != nil {
+		if resp.Response.StatusCode == http.StatusNotFound {
+			return "", fmt.Errorf("Error unable to find KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+		}
+		return "", fmt.Errorf("Error making Read request on KeyVault %q (Resource Group %q): %+v", vaultName, resourceGroup, err)
+	}
+
+	if resp.Properties == nil || resp.Properties.VaultURI == nil {
+		return "", fmt.Errorf("vault (%s) response properties or VaultURI is nil", keyVaultId)
+	}
+
+	return *resp.Properties.VaultURI, nil
+}
+
+// A KeyVaultId has the following structure: /subscriptions/<subscription_id>/resourceGroups/<resourceGroup_name>/providers/Microsoft.KeyVault/vaults/<vault_name>
+func getKeyVaultResourceGroupAndName(keyVaultId string) (string, string, error) {
+	if keyVaultId == "" {
+		return "", "", fmt.Errorf("keyVaultId is empty")
+	}
+
+	// Trim / from beginning and end and split KeyVaultId into pairs as structure suggests
+	components := strings.Split(strings.TrimSuffix(strings.TrimPrefix(keyVaultId, "/"), "/"), "/")
```

Comment:
`packer/builder/azure/common/client/resource.go` has a ` ParseResourceID` func that is used in Azure builders. That could make this func a bit cleaner.

Reply:
I was actually looking for something like this, let me have a look and will make amends!

Reply:
It turns out we hit a "import cycle not allowed" when using this package here, will dig further as the functionality needed is all there as you point out